p2p/src/lib.rs

```diff
@@ -117,24 +184,90 @@ where
         event: ConnectivityEvent<NetworkingBackend>,
     ) -> error::Result<()> {
         match event {
-            ConnectivityEvent::Accept(socket) => self.create_peer(socket, PeerRole::Inbound),
-            ConnectivityEvent::Connect(address) => self
-                .network
-                .connect(address)
-                .await
-                .map(|socket| self.create_peer(socket, PeerRole::Outbound))?,
+            ConnectivityEvent::Accept(peer_id, socket) => {
+                self.create_peer(peer_id, socket, PeerRole::Inbound)
+            }
+            ConnectivityEvent::Connect(address) => {
+                self.network.connect(address).await.map(|(peer_id, socket)| {
+                    self.create_peer(peer_id, socket, PeerRole::Outbound)
+                })?
+            }
         }
 
         Ok(())
     }
 
-    fn peer_discovered(&mut self, peers: &[NetworkingBackend::Address]) -> error::Result<()> {
-        println!("peers discovered: {:#?}", peers);
+    /// Try to establish new outbound connections if the total number of
+    /// active connections the local node has is below threshold
+    ///
+    // TODO: move all peer management to separate file
+    async fn auto_connect(&mut self) -> error::Result<()> {
+        // we have enough active connections
+        if self.peers.len() >= MAX_ACTIVE_CONNECTIONS {
+            return Ok(());
+        }
+
+        // we don't know of any peers
+        if self.discovered.is_empty() {
+            return Err(P2pError::NoPeers);
+        }
+
+        let npeers = std::cmp::min(
+            self.discovered.len(),
+            MAX_ACTIVE_CONNECTIONS - self.peers.len(),
+        );
+
+        // TODO: improve peer selection
+        let mut peers = Vec::new();
```

Comment:
Whenever you create a Vec::new(), planning to push into it, think that this can be exchanged with a mapping as long as the elements don't depend on each other.